### ISSUE TITLE
Monster targeting

### ISSUE THREAD

**Issue by EPuncker** [otland/forgottenserver#1223]
This is more like a question, do we have currently any way to make the following monster behaviors?

- target/attack lowest level player (pirates case)
- target/attack the player who is dealing the highest damage (Gnomevils case)
- target/attack the player wich X vocation and lowest level (Weeper case: target lowest lvl mage)

if there is no way to achieve it with current code, I'd like to suggest those features, otherwise just disregard this and tell me that this can be done easily and not resource intensive (something with onthink)

**Comment by EvanMC** [comment]
1. You could use onThink() (every 5 seconds maybe), Game.getSpectators(), loop through and find the lowest level player, then finally set target.
2. I would think there are two options for this one, onHealthChange() event and monitor all the incoming damage and find the highest damage dealer, or you could use Creature:getDamageMap().
3. Same with 1

**Comment by EPuncker** [comment]
@marksamman is that resource intensive or not?

and @EvanMC doesn't onCreatureAppear() works only when creature appear? and what if a knight uses exeta res, the intended behavior is that the monster turns again to the target player after 2 seconds again

**Comment by marksamman** [comment]
@EPuncker yes, it is resource intensive to call Game.getSpectators(), but it's the most efficient way to get creatures in an area, so use it when you have to. In this situation, you shouldn't have to, but you're limited to what you can do by the engine. It's unlikely to be the bottleneck of your server if it's only limited to certain monsters though. Marking this as an enhancement so that this can be customized more efficiently (perhaps an event called onSearchTarget, which lets you stop the default behavior and implement your own).

**Comment by gugahoa** [comment]
Cool idea about the event onSearchTarget, care do discuss it a little more? It would go under Events class?

I'm itching at trying my best to contribute to the repo. :(

**Comment by marksamman** [comment]
No, it would be a monster event, but it's marked as milestone 1.2 so even if you start working on it right now it won't be merged until after the release of 1.1.

**Comment by gugahoa** [comment]
Ok, in the future I'll try then. :(

**Comment by ranisalt** [comment]
@gugahoa come back and do this :P

**Comment by PrinterLUA** [comment]
I agree that monster AI should be more configureable.

**Comment by PrinterLUA** [comment]
I would like to make a pull request. What params should we include? 

function onSearchTarget(self,...?)

**Comment by joseluis2g** [comment]
onSearchTarget(self, searchType) ???

**Comment by ranisalt** [comment]
I don't think search type would be a parameter, as search type certainly can be decided on the function itself.

**Comment by joseluis2g** [comment]
what about targets id?

**Comment by PrinterLUA** [comment]
Also getting target list is possible with lua

**Comment by EPuncker** [comment]
hey everyone, long time huh? I just found out that someone did (almost) exactly what this bug report was made from, can't wait to test it, hope @marksamman take a look at the code 😛 

https://github.com/TwistedScorpio/OTHire/commit/8e567570eec5a2d4570a376bb7431ee280d36513
https://github.com/TwistedScorpio/OTHire/commit/5abc8c58073cd24ddb13d4b0c0dbc20a6f9e2221

**Comment by Znote** [comment]
or @EvilHero90

**Comment by EPuncker** [comment]
I know nobody is working on this, but there is another new thing I found that might be useful for monster behavior:
https://github.com/TwistedScorpio/OTHire/commit/38be35458290d7682f0a1c1053eeb288e449a781#diff-1cf6ea7bc7ef48f7c646a442e4e16a1f
